src/utils.ts

```diff
@@ -20,4 +43,60 @@ export async function waitForAnchor(doc: any): Promise<void> {
     }
     console.log(`anchor status reached for document ${doc.id.toString()}`)
     expect(doc.state.anchorStatus).toEqual(AnchorStatus.ANCHORED)
+}
+
+export async function buildIpfs(configObj): Promise<IpfsApi> {
+    let ipfs: IpfsApi
+    multiformats.multicodec.add(dagJose)
+    const format = legacy(multiformats, dagJose.name)
+    if (configObj.mode == "client") {
+        console.log(`Creating IPFS via http client, connected to ${configObj.apiURL}`)
+        ipfs = ipfsClient({url: configObj.apiURL, ipld: {formats: [format]}})
+    } else if (configObj.mode == "node") {
+        throw new Error("Creating in-process IPFS node is not currently supported")
+    } else if (configObj.mode == "none") {
+        return null
+    } else {
+        throw new Error(`IPFS mode "${configObj.mode}" not supported`)
+    }
+
+    return ipfs
+}
+
+export async function buildCeramic (configObj, ipfs: IpfsApi): Promise<CeramicApi> {
+    let ceramic
+    if (configObj.mode == "client") {
+        console.log(`Creating ceramic via http client, connected to ${configObj.apiURL}`)
+        ceramic = new CeramicClient(configObj.apiURL, { docSyncEnabled: true, docSyncInterval: 500 })
+    } else if (configObj.mode == "node") {
+        console.log("Creating ceramic local node")
+        const ceramicConfig: CeramicConfig = {
+            networkName: configObj.network,
+            ethereumRpcUrl: configObj.ethereumRpc,
+            anchorServiceUrl: configObj.anchorServiceAPI,
+        }
+        const [modules, params] = await Ceramic._processConfig(ipfs, ceramicConfig)
+        if (configObj.s3StateStoreBucketName) {
+            const s3StateStore = new S3StateStore(configObj.s3StateStoreBucketName)
+            modules.pinStoreFactory.setStateStore(s3StateStore)
+        }
+        ceramic = new Ceramic(modules, params)
+        await ceramic._init(true, true)
+    } else if (configObj.mode == "none") {
+        return null
+    } else {
+        throw new Error(`Ceramic mode "${configObj.mode}" not supported`)
+    }
+
+    const didProvider = new Ed25519Provider(seed)
+    await ceramic.setDIDProvider(didProvider)
+
+    return ceramic
+}
+
+export async function restartCeramic() {
```

Comment:
Yeah.  It doesn't really make sense to allow for ceramic2 to be restarted since there'd be no reason to ever run 2 in-process ceramic nodes (at that point it could be a unit test within the js-ceramic repo).  Since at least 1 of your ceramic instances will always be in `client` mode, and you can only restart a ceramic instance running in `node` mode, it seemed fine to have it be hard-coded which node it restarts.  Passing in the ceramic instance makes this more complicated as then you need to also map that instance to its corresponding config object, and reinstall it into the [`global`](https://github.com/3box/ceramic-integration-tests/blob/701a58b97c06b441407b4acf6c513475ef390aa7/src/index.ts#L8-L9) object correctly

Reply:
I did add some checks and comments to make this clearer, hopefully that helps

Reply:
Yeah I worry about 
> `ceramic2` is always expected to be in `client` mode.

There's no good guarantee for this and it's not obvious that it should be configured this way.
If we need such script guarantees mabye we should just not include an option to pass mode on ceramic?
I just think this is too error prone.

Reply:
Newest patch removed the `mode` option from the config file for `ceramic2`.  It's now only there for `ceramic`

Reply:
I agree this isn't super obvious though.  Man I wish JSON supported comments!

Reply:
So yeah my only concern here is that you have the proper context for how this works but from an outside perspective ceramic and ceramic2 seem like they should be interchangeable.

Even and even going over the code it’s not immediately clear that they are not.

Maybe it makes sense to call ceramic2 “ceramicClient”. Because I’m thinking what happens when we have ceramic3? And I guess that’s not relevant.

so “ceramic” can have modes but a “ceramicClient” can not.

Reply:
Yeah I agree.  I originally thought that ceramic and ceramic2 would indeed be totally interchangeable, but I think it's now becoming clear that that isn't the case and it's useful to have them take on more clearly defined roles within the tests.  How about `localCeramic` and `remoteCeramic`?

Reply:
Oh wait I guess that won't make much sense for the configuration where both are pointed at infra nodes to test interop between the infra nodes.  Okay `ceramic` and `ceramicClient` work then.